lib/spack/spack/modules/common.py

```diff
@@ -513,14 +551,35 @@ def environment_modifications(self):
         blacklist = self.conf.environment_blacklist
 
         # We may have tokens to substitute in environment commands
+
+        # Prepare a suitable transformation dictionary for the names
+        # of the environment variables. This means:
+        #
+        # 1. raising on invalid tokens
+        # 2. turn the valid tokens uppercase
+        #
+        transform = _token_invalidator(
+            'token {0} cannot be expanded in an environment variable name'
+        )
+        valid_tokens = (
+            'PACKAGE',
+            'VERSION',
+            'COMPILER',
+            'COMPILERNAME',
+            'COMPILERVER',
+            'ARCHITECTURE'
+        )
+        for token in valid_tokens:
+            transform[token] = str.upper
+
         for x in env:
-            x.name = self.spec.format(x.name)
+            x.name = self.spec.format(x.name, transform=transform)
             try:
                 # Not every command has a value
                 x.value = self.spec.format(x.value)
             except AttributeError:
                 pass
-            x.name = str(x.name).replace('-', '_').upper()
+            x.name = str(x.name).replace('-', '_')
```

Comment:
This appears to be the essential fix for https://github.com/LLNL/spack/issues/4684

Reply:
This is only half of the fix. Before everything was made uppercase, now I am making uppercase only the named tokens in the environment variable names. This is to ensure backward compatibility for the most common cases, like:
```yaml
modules:
  lmod:
    all:
      environment:
        set:
          '${PACKAGE}_ROOT': '${PREFIX}'
```
The main reason for that is that we have a lot of users relying on those variable being uppercase in their scripts. The argument is quite similar to the one made for `--run-tests` in #5132

Reply:
I missed here that module file generation may want to customize the extracted spec properties differently than other callers of `Spec.format` so I can see why it is useful to pass a dictionary of transformation functions. Originally I was confused why you didn't just pass in a set of Spec properties that should be supported. A dictionary of functions makes sense to me for this. That being said I mention elsewhere that I think this dictionary shouldn't contain functions for Spec properties you don't support (for a particular invocation of `Spec.format`).